morpho/processors/diagnostics/CalibrationProcessor.py

```diff
@@ -0,0 +1,278 @@
+        root_post_tree: Tree containing analysis posteriors. Defaults to "analysis".
+        post_param_names: List of strings naming posteriors produced by Stan analysis for a parameters of interest. Defaults to self.in_param_names.
+        verbose: If True, print information for each pseudo-experiment and a summary at the end. If False, only print the summary.
+        quantile: If True, compute quantile credible intervals. Otherwise, compute highest density intervals.
+        check_if_nonzero: If True, check whether posteriors allow the parameters to be distinguished from zero (given some credible interval).
+        
+    Results:
+        coverages: dictionary containing coverage of interval given by self.cred_interval, for each parameter in self.in_param_names
+    '''
+    def InternalConfigure(self,params):
+        #Required input
+        self.files = reader.read_param(params,'files','required')
+        self.in_param_names = reader.read_param(params,'in_param_names','required')
+        
+        #Optional input
+        self.cred_interval = reader.read_param(params,'cred_interval',[0.05, 0.95])
+        self.root_in_tree = reader.read_param(params,'root_in_tree','input')
+        self.root_post_tree = reader.read_param(params,'root_post_tree','analysis')
+        self.post_param_names = reader.read_param(params,'post_param_names',self.in_param_names)
+        self.verbose = reader.read_param(params,'verbose',True)
+        self.quantile = reader.read_param(params,'quantile',False)
+        self.check_if_nonzero = reader.read_param(params,'check_if_nonzero',False)
+        
+        #Other
+        self.failed_runs = []
+        return True
+    
+    
+    def perform_calibration(self):
+        logger.info("Calibrating credible interval results")
+        
+        #Defining credibility
+        if len(self.cred_interval)==1:
+            alpha = self.cred_interval[0]
+        elif len(self.cred_interval)==2:
+            alpha = self.cred_interval[1]-self.cred_interval[0]
+        else:
+            logger.error("Input a credible interval list with one or two elements.")
+            return
+        
+        #Setting up variables before loop
+        calib_bounds = {name:[] for name in self.in_param_names}
+        consistent_with_zero = {name:0 for name in self.in_param_names}
+        n_recovered_inputs = {name:0 for name in self.in_param_names}
+        #Dictionary to keep track of sums of quantities, so that averages can be taken after the loop
+        sums = {name:dict.fromkeys(['median', 'mean', 'lower', 'upper'],0) for name in self.in_param_names}
+        
+        for i in range(len(self.files)):
+            #Reading input values and posteriors from root files
+            try:
+                input_vals, posterior_arrays = self._load_inputs_and_posteriors(self.files[i])
+            except AttributeError as error:
+                logger.warning(error)
+                self.failed_runs.append(filename)
```

Comment:
`filename -> self.files[i]` , right?
Does it work this way?

Reply:
Sorry for the inconsistency. Do you think it would be better to actually use `self.files[i]` throughout? (Except for in the` _load_inputs_and_posteriors` function itself.)

Reply:
Sorry for the inconsistency. I see what you did below with enumerate; that's perfect.